pkg/kubelet/token/token_manager.go

```diff
@@ -0,0 +1,118 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package token
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+func NewManager(c corev1.CoreV1Interface) *Manager {
+	return &Manager{
+		getToken: func(name, namespace string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+			return c.ServiceAccounts(namespace).CreateToken(name, tr)
+		},
+		cache: make(map[string]*authenticationv1.TokenRequest),
+		now:   time.Now,
+	}
+}
+
+type Manager struct {
+	getToken func(name, namespace string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error)
+
+	// cmu guards the cache
+	cmu   sync.RWMutex
+	cache map[string]*authenticationv1.TokenRequest
+
+	// mocked for testing
+	now func() time.Time
+}
+
+func (m *Manager) GetServiceAccountToken(name, namespace string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+	key := keyFunc(name, namespace, tr)
+	ctr, ok := m.get(key)
+
+	var err error
+	if !ok || m.requiresRefresh(ctr) {
+		tr, err := m.getToken(name, namespace, tr)
+		if err != nil {
+			glog.Errorf("couldn't update token %s: %v", key, err)
+		} else {
+			m.set(key, tr)
+		}
+	}
+
+	ctr, ok = m.get(key)
+	if !ok && err != nil {
+		return nil, fmt.Errorf("failed to fetch initial token: %v", key, err)
+	}
+	if !ok {
+		// wat
+		return nil, fmt.Errorf("failed to fetch token")
+	}
+	if m.expired(ctr) && err != nil {
+		return nil, fmt.Errorf("token %s expired and refresh failed: %v", key, err)
+	}
+	if m.expired(ctr) {
+		// wat
+		return nil, fmt.Errorf("token %s expired but refresh succeeded", key)
+	}
+	return ctr, nil
+}
+
+func (m *Manager) get(key string) (*authenticationv1.TokenRequest, bool) {
+	m.cmu.RLock()
+	defer m.cmu.RUnlock()
+	ctr, ok := m.cache[key]
+	return ctr, ok
+}
+
+func (m *Manager) set(key string, tr *authenticationv1.TokenRequest) {
+	m.cmu.Lock()
+	defer m.cmu.Unlock()
+	m.cache[key] = tr
+}
+
+func (m *Manager) expired(t *authenticationv1.TokenRequest) bool {
+	return m.now().After(t.Status.ExpirationTimestamp.Time)
+}
+
+// requiresRefresh returns true if the token is older than 80% of its total
+// ttl, or if the token is older than 24 hours.
+func (m *Manager) requiresRefresh(tr *authenticationv1.TokenRequest) bool {
+	now := m.now()
+	exp := tr.Status.ExpirationTimestamp.Time
+	iat := exp.Add(-1 * time.Duration(*tr.Spec.ExpirationSeconds) * time.Second)
```

Comment:
could cause nil pointer dereference here

Reply:
This should be "guaranteed" to be non-nil because the ServiceAccountVolumeProjection will have already been defaulted. We could be defensive, return false and log an error?

Reply:
Yeah, better safe than sorry. I don't think static pods go through defaulting.

Reply:
Done.